### ISSUE TITLE
False positive: function-redefined triggered for local functions with same name inside different branches

### ISSUE THREAD

**Issue by Akshay9715** [pylint-dev/pylint#10622]
### Bug description

After recent improvements to the `function-redefined` checker (PR #9894 ), the Primer run for the **Sentry** repository revealed a potential **false positive**.

The file:
[`sentry/src/sentry/workflow_engine/processors/detector.py`](https://github.com/getsentry/sentry/blob/e6c0e56db3f6a248cbc94f783ec32565c291526c/src/sentry/workflow_engine/processors/detector.py)
contains two *local* function definitions named `_extract_events_lookup_key`, defined inside **different branches** of the same outer function:

```python
def get_detectors_by_groupevents_bulk(event_list: list[GroupEvent]) -> Mapping[str, Detector]:
    ...
    if events_with_occurrences:
        def _extract_events_lookup_key(detector: Detector) -> int:
            return detector.id
        ...

    if error_events:
        def _extract_events_lookup_key(detector: Detector) -> int:
            return detector.project_id
        ...
```

Both definitions are valid Python — they are **scoped within their respective conditional branches** and do not overlap in execution.
However, Pylint currently reports:
```python
function-redefined: function already defined line 139
```

This indicates that the `function-redefined` checker does not yet distinguish between **local function redefinitions in separate branches** versus actual redefinitions within the same scope.

✅ Expected behavior

Pylint should not emit a `function-redefined` message when the same function name is defined in **separate, mutually exclusive code branches** inside another function.

🧪 Steps to reproduce
```python
def f(x):
    if x:
        def g(): pass
    else:
        def g(): pass
```

Current output:
```
test.py:5:4: E0102: function already defined line 3 (function-redefined)
```

**Expected output:**
No `function-redefined` message — since these are distinct local definitions in different control flow paths.

**⚙️ Pylint version**
```
pylint 4.0.0-dev0
Python 3.12 / 3.13
```

**💬 Notes**

* This false positive existed before but was previously hidden by underscored-function filtering.
* The recent PR exposed it correctly during the Primer for Sentry.
* Suggested fix: improve scope tracking in the redefinition checker to account for **branch-local function definitions.**

Related discussion: https://github.com/pylint-dev/pylint/pull/10606

**Comment by jacobtylerwalls** [comment]
How do we know the assignments are exclusive?

**Comment by jacobtylerwalls** [comment]
The sentry example is different from the minimal example.

**Comment by Akshay9715** [comment]
@jacobtylerwalls
In the minimal example I included, the two local functions are defined inside an `if/else` structure, so only one branch executes at runtime — they are mutually exclusive.

In the Sentry example, the two `_extract_events_lookup_key` functions are inside separate `if` statements (`if events_with_occurrences:` and `if error_events:`). You’re right that, in some cases, both branches could potentially execute, which would indeed redefine the function at runtime.

**Comment by Akshay9715** [comment]
### ✅ Reproduction check for function-redefined false positive

I tested minimal examples to verify the behavior of Pylint regarding local function redefinitions:

1️⃣ Mutually exclusive branches (if/else)
```python
def f(x):
    if x:
        def g(): pass
    else:
        def g(): pass
```

Result (Pylint 4.0.0-dev0):
* No E0102 warning — correct behavior.
* Confirms that false positives in mutually exclusive branches are no longer present in this version.

2️⃣ Separate if statements (non-exclusive)
```python
def f(x):
    if x == 5:
        def g(): pass
    if x % 2 == 1:
        def g(): pass
```

Result (Pylint 4.0.0-dev0):
```
test_func_redef.py:5:8: E0102: function already defined line 3 (function-redefined)
```

This is expected, as both definitions can execute, causing a real redefinition.

3️⃣ Mutually exclusive if/elif/else (slightly more complex)
```python
def f(x, y):
    if x:
        def g(): pass
    elif y:
        def g(): pass
    else:
        def g(): pass
```
Result (Pylint 4.0.0-dev0):
* No E0102 warning — correct behavior.

4️⃣ Environment
```
pylint 4.0.0-dev0
astroid 4.1.0-dev0
Python 3.13.1 (tags/v3.13.1)
```

4️⃣ Additional context:

The false positive existed in older Pylint versions for mutually exclusive branches (if/else), but it appears fixed in the dev version I tested.

The checker correctly warns for actual redefinitions in non-exclusive branches.

5️⃣ Conclusion:

* Mutually exclusive branches: ✅ correctly handled, false positive gone.
* Non-exclusive branches: ✅ correctly flagged.
* The false positive for mutually exclusive branches (if/else) is not present in this version — Pylint correctly handles that case.

Should this issue now be closed as resolved in the latest dev version, or do you want me to propose a PR to further refine the checker for **more complex mutually exclusive branches (e.g., nested branches, loops)?**

**Comment by DanielNoord** [comment]
Likely fixed becuase of the new constraint logic, let's close! Thanks for the clear reproduction analysis.
